File: frontend/js/views/asset_disposal.js
// Fixed-asset disposal generator — gain/loss with §1245 ordinary recapture and
// §1231 split, via /calc/asset-disposal.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
let LAST_DOC = null;

export async function renderAssetDisposal(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.dispose.h1.title">// FIXED-ASSET DISPOSAL</span></h1>
        <p class="muted small" data-i18n="view.dispose.hint.intro">
            The gain or loss when a business sells or scraps a depreciable asset. Net book value is cost less
            accumulated depreciation; gain or loss is proceeds less net book value. For §1245 property, gain is
            ordinary income to the extent of prior depreciation (recapture), any amount above the original cost
            is §1231 capital gain, and a sale below book value is an ordinary §1231 loss. Drafting aid, not tax advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.dispose.h2.inputs">Disposal inputs</h2>
            <form id="dispose-form" class="inline-form">
                <label><span data-i18n="view.dispose.label.company">Company</span>
                    <input type="text" name="company_name" value="Acme Co" required></label>
                <label><span data-i18n="view.dispose.label.asset">Asset</span>
                    <input type="text" name="asset_label" value="CNC machine" required></label>
                <label><span data-i18n="view.dispose.label.cost">Original cost ($)</span>
                    <input type="number" step="100" min="0" name="cost_usd" value="10000" required></label>
                <label><span data-i18n="view.dispose.label.accum">Accumulated depreciation ($)</span>
                    <input type="number" step="100" min="0" name="accumulated_depreciation_usd" value="6000" required></label>
                <label><span data-i18n="view.dispose.label.proceeds">Sale proceeds ($)</span>
                    <input type="number" step="100" min="0" name="proceeds_usd" value="7000"></label>
                <label><span data-i18n="view.dispose.label.disposaldate">Disposal date</span>
                    <input type="date" name="disposal_date" value="2026-06-15" required></label>
                <label><span data-i18n="view.dispose.label.date">Statement date</span>
                    <input type="date" name="date" value="2026-07-01" required></label>
                <label><span data-i18n="view.dispose.label.note">Note (optional)</span>
                    <input type="text" name="note" value="" placeholder="${esc(t('view.dispose.ph.note'))}"></label>
            </form>
        </div>
        <div id="dispose-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#dispose-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            company_name: (fd.get('company_name') || '').trim(),
            asset_label: (fd.get('asset_label') || '').trim(),
            cost_usd: Number(fd.get('cost_usd')) || 0,
            accumulated_depreciation_usd: Number(fd.get('accumulated_depreciation_usd')) || 0,
            proceeds_usd: Number(fd.get('proceeds_usd')) || 0,
            disposal_date: fd.get('disposal_date'),
            date: fd.get('date'),
            note: (fd.get('note') || '').trim(),
        };
        try {
            const doc = await api.calcAssetDisposal(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.dispose.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase()];
    if (doc.statutory_citation) lines.push(doc.statutory_citation);
    lines.push('');
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const glCls = doc.gain_loss_usd >= 0 ? 'pos' : 'neg';
    const glLabel = doc.is_loss ? t('view.dispose.card.loss') : t('view.dispose.card.gain');
    const el = mount.querySelector('#dispose-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card ${glCls}"><div class="label">${esc(glLabel)}</div>
                    <div class="value">${money(Math.abs(doc.gain_loss_usd))}</div></div>
                <div class="card"><div class="label" data-i18n="view.dispose.card.book">Net book value</div>
                    <div class="value">${money(doc.net_book_value_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.dispose.card.ordinary">Ordinary (§1245)</div>
                    <div class="value">${money(doc.ordinary_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.dispose.card.capital">§1231 capital gain</div>
                    <div class="value">${money(doc.section_1231_gain_usd)}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="dispose-copy" type="button" data-i18n="view.dispose.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="dispose-download" type="button" data-i18n="view.dispose.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#dispose-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.dispose.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.dispose.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#dispose-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'asset-disposal.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/asset_disposal.rs
//! Fixed-asset disposal (gain/loss) — the gain or loss when a business sells or
//! scraps a depreciable asset, with the §1245 ordinary-income recapture split.
//! Net book value is cost less accumulated depreciation; gain or loss is proceeds
//! less net book value. For §1245 personal property, gain is ordinary income to
//! the extent of prior depreciation (recapture); any amount above the original
//! cost is §1231 (capital) gain; a sale below book value is an ordinary §1231
//! loss. Distinct from the rental §1250 recapture module, which recaptures only
//! depreciation above straight-line. Drafting aid, not tax advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct AssetDisposalInput {
    pub company_name: String,
    pub asset_label: String,
    /// Original capitalized cost.
    pub cost_usd: f64,
    /// Accumulated depreciation taken to the disposal date.
    pub accumulated_depreciation_usd: f64,
    /// Sale proceeds (0 for a scrap/abandonment).
    #[serde(default)]
    pub proceeds_usd: f64,
    pub disposal_date: String,
    pub date: String,
    #[serde(default)]
    pub note: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct AssetDisposal {
    pub title: String,
    /// Cost − accumulated depreciation.
    pub net_book_value_usd: f64,
    /// Proceeds − net book value (negative = loss).
    pub gain_loss_usd: f64,
    /// Ordinary income: §1245 depreciation recapture (or an ordinary §1231 loss).
    pub ordinary_usd: f64,
    /// §1231 (capital) gain — the portion of gain above original cost.
    pub section_1231_gain_usd: f64,
    /// True when the disposal produced a loss.
    pub is_loss: bool,
    pub statutory_citation: String,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &AssetDisposalInput) -> AssetDisposal {
    let book = cents(i.cost_usd - i.accumulated_depreciation_usd);
    let gain_loss = cents(i.proceeds_usd - book);
    let is_loss = gain_loss < 0.0;

    let (ordinary, sec_1231) = if gain_loss > 0.0 {
        // §1231 capital gain only for proceeds above original cost; the rest of
        // the gain is ordinary §1245 recapture.
        let capital = cents((i.proceeds_usd - i.cost_usd).max(0.0));
        (cents(gain_loss - capital), capital)
    } else {
        // A loss (or zero) on §1231 property is an ordinary loss.
        (gain_loss, 0.0)
    };

    let outcome = if gain_loss > 0.0 {
        format!(
            "The disposal produces a gain of {}: {} is ordinary income (§1245 depreciation recapture) and {} is §1231 capital gain.",
            money(gain_loss), money(ordinary), money(sec_1231)
        )
    } else if is_loss {
        format!("The disposal produces an ordinary §1231 loss of {}.", money(gain_loss.abs()))
    } else {
        "The asset is disposed of at its net book value, with no gain or loss.".to_string()
    };

    let calc_body = format!(
        "Net book value is {} (cost {} less accumulated depreciation {}). Proceeds are {}, so the gain/loss is {}. {}",
        money(book),
        money(i.cost_usd),
        money(i.accumulated_depreciation_usd),
        money(i.proceeds_usd),
        money(gain_loss),
        outcome
    );

    let mut clauses = vec![
        DocClause {
            heading: "Asset".into(),
            body: format!(
                "Company: {}\nAsset: {}\nDisposal date: {}\nStatement date: {}",
                i.company_name, i.asset_label, i.disposal_date, i.date
            ),
        },
        DocClause { heading: "Disposal".into(), body: calc_body },
        DocClause {
            heading: "Journal Entry".into(),
            body: format!(
                "Remove the asset: credit the asset {} and debit accumulated depreciation {}. Debit cash {}. {}",
                money(i.cost_usd),
                money(i.accumulated_depreciation_usd),
                money(i.proceeds_usd),
                if gain_loss > 0.0 {
                    format!("Credit gain on disposal {}.", money(gain_loss))
                } else if is_loss {
                    format!("Debit loss on disposal {}.", money(gain_loss.abs()))
                } else {
                    "No gain or loss.".to_string()
                }
            ),
        },
    ];

    let note = i.note.trim();
    if !note.is_empty() {
        clauses.push(DocClause { heading: "Note".into(), body: note.to_string() });
    }

    AssetDisposal {
        title: "Fixed-Asset Disposal — Gain/Loss".into(),
        net_book_value_usd: book,
        gain_loss_usd: gain_loss,
        ordinary_usd: ordinary,
        section_1231_gain_usd: sec_1231,
        is_loss,
        statutory_citation: "IRC §1245 / §1231".to_string(),
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> AssetDisposalInput {
        AssetDisposalInput {
            company_name: "Acme Co".into(),
            asset_label: "CNC machine".into(),
            cost_usd: 10_000.0,
            accumulated_depreciation_usd: 6_000.0,
            proceeds_usd: 7_000.0,
            disposal_date: "2026-06-15".into(),
            date: "2026-07-01".into(),
            note: String::new(),
        }
    }

    #[test]
    fn gain_below_cost_all_ordinary() {
        let d = generate(&base());
        assert!(close(d.net_book_value_usd, 4_000.0));
        assert!(close(d.gain_loss_usd, 3_000.0));
        assert!(close(d.ordinary_usd, 3_000.0));
        assert!(close(d.section_1231_gain_usd, 0.0));
        assert!(!d.is_loss);
    }

    #[test]
    fn gain_above_cost_splits_1231() {
        let d = generate(&AssetDisposalInput { proceeds_usd: 12_000.0, ..base() });
        assert!(close(d.gain_loss_usd, 8_000.0));
        // Recapture capped at accumulated depreciation; excess above cost is §1231.
        assert!(close(d.ordinary_usd, 6_000.0));
        assert!(close(d.section_1231_gain_usd, 2_000.0));
    }

    #[test]
    fn sold_below_book_is_loss() {
        let d = generate(&AssetDisposalInput { proceeds_usd: 3_000.0, ..base() });
        assert!(close(d.gain_loss_usd, -1_000.0));
        assert!(close(d.ordinary_usd, -1_000.0));
        assert!(d.is_loss);
    }

    #[test]
    fn at_book_no_gain_loss() {
        let d = generate(&AssetDisposalInput { proceeds_usd: 4_000.0, ..base() });
        assert!(close(d.gain_loss_usd, 0.0));
        assert!(!d.is_loss);
    }

    #[test]
    fn scrap_full_recapture_then_book_loss() {
        // Proceeds 0 on a fully-depreciated-down asset → loss equal to book value.
        let d = generate(&AssetDisposalInput { proceeds_usd: 0.0, ..base() });
        assert!(close(d.gain_loss_usd, -4_000.0));
        assert!(d.is_loss);
    }

    #[test]
    fn citation_present() {
        let d = generate(&base());
        assert_eq!(d.statutory_citation, "IRC §1245 / §1231");
    }
}
